### PR TITLE
[bugfix]: preserve FSDP hooks for RMSNorm qk norms

### LINKED ISSUE #1379: RMSNorm.forward_native doesn't unwrap DTensor weight under FSDP CPU offload
## Background

This issue tracks an FSDP CPU-offload edge case in `fastvideo/layers/layernorm.py`'s `RMSNorm.forward_native` that was originally bundled into PR #572 by @Edenzzzz as a workaround, but was dropped from that PR's scope to keep it focused on distributed weight loading. The underlying bug still exists; this issue tracks it so it doesn't get lost.

## The bug

When a model uses FSDP with CPU offload, `RMSNorm.forward_native`'s `self.weight` can remain as a `DTensor` rather than being unwrapped to a local tensor, which then breaks downstream ops. The author's own inline note (from PR #572's original branch state `c2b66087d`) flagged it as a known FSDP workaround that needs upstream investigation:

```python
# TODO(wenxuan): When using CPU offload, FSDP has a bug that
# doesn't unwrap DTensor in final_layer_norm. Report this
weight = self.weight.to_local().to(x.device) if isinstance(self.weight, DTensor) else self.weight
```

## Workaround (saved on the original PR #572 branch)

A 5-line addition to `RMSNorm.forward_native` immediately before `x = x * weight`:

```python
weight = self.weight.to_local().to(x.device) if isinstance(self.weight, DTensor) else self.weight
```

(See PR #572's pre-drop commit `c2b66087d` for the original FastVideo patch; commit `b7c338210` is the scope-drop that removed it before merge.)

## What this issue is asking for

Before re-landing the workaround as a separate PR, we need:

- [ ] **A reliable reproducer**: what FSDP version, model, hardware, and code path triggers this? The original PR didn't ship a regression test.
- [ ] **An upstream issue link** (pytorch/pytorch?): is this a known FSDP-side bug, or specific to how FastVideo wires FSDP + RMSNorm? If it's pytorch-side, the workaround should be retired when upstream lands a fix.
- [ ] **A unit test** that exercises the CPU-offload + DTensor weight path on `RMSNorm.forward_native` so the workaround can ship with a regression gate.
- [ ] **A clear behavioral spec**: does the bug surface only on `final_layer_norm` (per the comment) or anywhere `RMSNorm.forward_native` runs with `has_weight=True` under FSDP CPU offload? The comment says the former; needs verification.

## What this issue is NOT

- This is **not** about `RMSNorm.forward_cuda` — only the `forward_native` path.
- This is **not** blocking PR #572 (already merged at `af2ee9c78` without the workaround).
- This is **not** asking for the workaround to be merged blindly; a reproducer + test + upstream link first.

## Suggested labels

`bug`, `help wanted`, `scope: model`, `scope: distributed`.

---

*This issue was filed by one of @SolitaryThinker's AI reviewer/fix agents (Gob) to track a known issue dropped from PR #572. Ping @SolitaryThinker or @Edenzzzz if any of the framing is off.*

### PR DESCRIPTION
**Pull request by macthecadillac**
## Summary

Fixes #1379.

This replaces the remaining direct q/k RMSNorm `forward_native(...)` calls with normal module dispatch so FSDP hooks run:

- `self.norm_q.forward_native(query)` -> `self.norm_q(query)`
- `self.norm_k.forward_native(key)` -> `self.norm_k(key)`

The remaining production call sites were in causal Wan and MatrixGame2 causal transformer blocks. This keeps the selected RMSNorm implementation unchanged while avoiding bypassing `nn.Module.__call__` under FSDP.

## Root Cause

Calling `RMSNorm.forward_native(...)` directly bypasses FSDP2 module hooks. With sharded RMSNorm weights, this can leave `self.weight` as a `DTensor`, and the native multiply fails with a mixed Tensor/DTensor error. Calling `RMSNorm(x)` lets FSDP unwrap/localize parameters as expected.

## Issue #1379 Investigation Checklist

- [x] Reliable reproducer: added `fastvideo/tests/layers/test_rmsnorm_forward_dispatch.py` and reproduced on Modal 2x L40S with PyTorch `2.11.0+cu128`, FSDP2 `fully_shard`, and both regular sharding and `CPUOffloadPolicy`.
- [x] Upstream issue determination: no exact matching PyTorch issue was found. The repro shows `RMSNorm(x)` works while direct `RMSNorm.forward_native(x)` fails, so this appears to be FastVideo bypassing FSDP module hooks rather than an FSDP-side bug.
- [x] Unit test for CPU-offload + DTensor weight path: the new test covers `direct_cpu_offload`, where `norm.weight` remains a CPU `DTensor` and direct `forward_native` fails with the mixed Tensor/DTensor error; it also verifies module dispatch succeeds with CPU offload.
- [x] Behavioral spec: not final-layer-only. The failure can affect any FSDP-sharded `RMSNorm(has_weight=True)` called directly through `forward_native`; it reproduces with and without CPU offload.

## Tests

- `uvx pre-commit==4.0.1 run --all-files`
  - passed after the review follow-up commit
- Modal L40S: `pytest fastvideo/tests/layers/test_rmsnorm_forward_dispatch.py -vs`
  - 2x L40S, latest branch includes the 120-second `torchrun` timeout follow-up
  - `5 passed, 14 warnings in 38.28s`
- Modal L40S: `pytest fastvideo/tests/transformers/test_wanvideo.py -vs`
  - 1x L40S, production fix applied
  - `1 passed, 14 warnings in 50.05s`
- Modal L40S targeted Wan T2V SSIM:
  - `FASTVIDEO_SSIM_MODEL_ID=Wan2.1-T2V-1.3B-Diffusers pytest fastvideo/tests/ssim/test_wan_t2v_similarity.py -vs`
  - 2x L40S, commit `75e1659310ae306d342b6bfcd0e8ef7ebb0b5caa`
  - `2 passed, 18 warnings in 123.41s`
  - `FLASH_ATTN`-parametrized mean SSIM: `0.976557461420695`
  - `TORCH_SDPA` mean SSIM: `0.9821627881791857`
  - Note: the Modal image's FlashAttention import failed and the `FLASH_ATTN`-parametrized run fell back to Torch SDPA.

## Checklist

- [x] I ran pre-commit run --all-files and fixed all issues
- [x] I added or updated tests for my changes
- [x] I updated documentation if needed
- [x] I considered GPU memory impact of my changes

For model/pipeline changes, also check:

- [x] I verified targeted Wan T2V SSIM regression tests pass on L40S
- [x] I updated the support matrix if adding a new model

Documentation/support-matrix notes: no docs or support matrix update was needed; this does not add a model or user-facing API. GPU memory impact should be neutral because this only routes existing RMSNorm calls through module dispatch so FSDP hooks can run.

### CODE REVIEW

File: fastvideo/tests/layers/test_rmsnorm_forward_dispatch.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    try:
+        mesh = init_device_mesh("cuda", (world_size,))
+        norm = RMSNorm(HIDDEN_SIZE, eps=1e-6, has_weight=True).to(device)
+        with torch.no_grad():
+            norm.weight.fill_(1.0)
+
+        fsdp_kwargs: dict[str, Any] = {"mesh": mesh}
+        if mode.endswith("cpu_offload"):
+            fsdp_kwargs["offload_policy"] = CPUOffloadPolicy(pin_memory=False)
+        sharded_norm = fully_shard(norm, **fsdp_kwargs)
```
Comment (SolitaryThinker) [inline fastvideo/tests/layers/test_rmsnorm_forward_dispatch.py]: **P2: exercise the production FSDP ownership boundary.** This directly `fully_shard`s the RMSNorm, but FastVideo's Wan/MatrixGame2 configs shard each `blocks.<n>` parent. The block's pre-forward hook therefore all-gathers descendant parameters before its forward body, so the old `norm_q.forward_native()` runs with an already-unsharded weight. This setup proves that direct methods bypass hooks when RMSNorm itself is the FSDP unit; it does not reproduce either edited call site. Please make the regression wrap and call the affected block (or a minimal parent owning a child RMSNorm) with the production topology and show that old code fails while the new code passes. If the old code passes, this should be framed as defensive cleanup rather than closing #1379.

**Review by SolitaryThinker — COMMENTED** [review]
Reviewed the production dispatch changes and regression coverage. One inline correctness finding; CI status intentionally excluded.

**Review by SolitaryThinker — APPROVED** [review]
